caip-browser_extension_provider_communication.md

```diff
@@ -0,0 +1,137 @@
+
+### Message format
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "/caip-x/schemas/provider-request.schema.json",
+  "title": "Provider / Web Extension message",
+  "description": "A request sent between a provider and a web extension.",
+  "type": "object",
+  "properties": {
+    "type": {
+      "description": "The message type, used to identify this as a CAIP-X message",
+      "const": "caip-x"
+    },
+    "data": {
+      "description": "A JSON-RPC message",
+      "type": "object"
+    }
+  },
+  "required": ["type", "data"]
+}
+```
+
+### `externally_connectable`
+
+A web extension can use the `externally_connectable` manifest field to accept messages from websites and other extensions. This permission can be configured to limit which sites and other extensions can send messages to the web extension. How this permission is configured is out-of-scope for this proposal; this proposal only concerns sites that the web extension allows messages from.
+
+
+The web extension can:
+* handle a connection from the website or other extensions by using `chrome.runtime.onConnectExternal.addListener((port) => {...})`
+* send messages by using `port.postMessage()`
+* receive messages by using `port.onMessage.addListener()`
+  * incoming messages should be validated according to the message format specification above
+
+The webpage embedded provider can:
+* initiate a connection with the web extension by using `port = browser.runtime.connect()`
+* send messages by using `port.postMessage()`
+* receive messages by using `port.onMessage.addListener()`
+
+### Caveats
+Currently Firefox does not support `externally_connectable` yet, but they are [considering implementing it](https://bugzilla.mozilla.org/show_bug.cgi?id=1319168). Meanwhile, extension wallets on Firefox will need to continue injecting an inpage provider for the dApp.
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+While web extensions could realize the benefits of `externally_connectable` without using it with a standardized interface, this would mean that every web extension would have to ship a provider library that worked specifically for their interface and that Dapps would need to import them. It wouldn't be feasible for Dapps to import every single web extension's specific provider implementation, making this approach less viable.
```

Comment:
This seems to be a rationale for "having a standard" versus doing the same thing without a standard. That seems atypical. I would expect the rationale section to discuss why we're doing this in the first place, not why we're standardizing it.

Reply:
The point made here about inter-operability is totally relevant, but maybe we could frame it more like:

* `externally_connectable` allows websites to communicate with wallets without requiring code injection
* This message format allows us to disambiguate these messages from others that extensions and websites might be sending on this same interface
* Using the standard as the message type enables compatibility with any website <=> wallet message formats (such as EIP-1193, CAIP-25, CAIP-27, etc.), including future formats
* Standardizing this allows websites and wallets to be inter-operable in the same way they are today with EIP-1193.

Reply:
Maybe this is an appropriate place to list the benefits of not injecting code as well, which I see as:
* Reduced extension permissions
* Improved performance
* Improved website compatibility

Maybe also worth explaining how this reduces complexity for dapp developers? I don't know how to express this idea, but it's the part about how we're forcibly injecting code into another persons application, and they need to consider this in how they build and design that application, and when debugging and monitoring and reviewing bug reports and so on. e.g. like some people are out there triaging Sentry reports for errors that supposedly come from their app, but it's our code. It's a mess for website authors, and this simplifies it somewhat.